### ISSUE TITLE
Translate Odds anonymous functions and expressions

### ISSUE THREAD

**Issue by dannyech** [alexandramedway/odds#42]
Translate all Odds function calls with anonymous expressions to Python function calls in which the expressions are evaluated outside the function call, assigned to variables, and then passed as variables

Ok! Figured out one way to translate Odds function calls that have anonymous functions as arguments to Python.

In Odds, say we have:
```
do foo = (bar) -> return bar()
do str = foo( () -> return "Hello, World" )
```
We **cannot** translate this to
```python
def foo_0(bar_1):
    return bar_1()
str_2 = foo_0(def anon_3(): 
                      return "Hello, World")
```

because you cannot create functions inside function calls.

This means that all **None ID** expression must be first evaluated outside the python function, assigned a variable, and then passed. We write the above code as:
```python
def foo_0(bar_1):
    return bar_1()

def anon_2(): 
    return "Hello, World"
str_3 = foo_0(anon_2)
```
Even arguments that are not functions but still **None ID** expressions must be evaluated outside of functions calls because the generator cannot distinguish between anonymous functions declared in a function call and anonymous functions declared outside a function call. So the Odds code:
```
do add = (num) -> return num + 1
do sum = add(2 * 2)
```
needs top be translated to:
```python
def add_0(num_1):
    return num_1 + 1

anon_2 = 2 * 2
sum = add_0(anon_2)
```

**Comment by dannyech** [comment]
Moreover, the same needs to be done with return statements!

Say we have in Odds:
```
do alex = () -> return () -> return "Was up late last night"
```
Would need to be translated as:
```python
def alex_0():
    def anon_1():
        anon_2 = "was up late last night"
        return anon_2
   return anon_1
```

**Comment by alexandramedway** [comment]
Why do we need to do this with return statements?

**Comment by dannyech** [comment]
@ Alexandra. You can't write in Python:

```python
def example():
    return def some_function():
                   return "Some String"
```

The two Python examples in my original post would actually be:

```python
def foo_0(bar_1):
    return bar_1()

def anon_2(): 
    anon_3 = "Hello, World"
    return anon_3
str_4 = foo_0(anon_2)
```
and
```python
def add_0(num_1):
    anon_2 = num_1 + 1
    return anon_2

anon_3 = 2 * 2
sum = add_0(anon_3)
```

Basically, we can only pass variable names and not expressions.

**Comment by akalicki** [comment]
I get why you have to do it for anonymous functions (and this looks great for those), but why for normal expressions? I see no issue with
```python
def add_0(num_1):
    return num_1 + 1

sum = add_0(2*2)
```
which gets parsed to `add_0(Binop(2, Mult, 2))`. Or:

```python
def foo_0(bar_1):
    return bar_1()

def anon_2(): 
    return "Hello, World"

str_4 = foo_0(anon_2)
```

**Comment by dannyech** [comment]
We don't need to per se, but it might be easier to code it that way. We will talk it over on Sunday.

**Comment by akalicki** [comment]
This will be addressed in translating our SAST (#51) to a PAST (#52) and the subsequent code generation (#53). After talking to Edwards we decided to go with the moving anonymous functions up one level to become "real" functions. This will be handled in the move from the SAST to the PAST so code generation treats them as normal functions.
